File: frontend/src/types/memory.ts
/**
 * 长期记忆管理相关类型定义
 *
 * 对应后端 /memory 系列 API(app/routers/memory.py):
 * - User Profile (1:1,结构化字段 + 自由文本):影响 user_agent 评判标准与 checklist 生成
 * - 全局长期记忆(1:1,自由文本):跨项目通用经验,注入 user_agent
 * - 分项目记忆(1:N,按 repo_url 聚合):注入 react_agent,影响审计方向
 *
 * 三类记忆均可由用户手动编辑;分项目记忆还会在任务完成时由 agent 自动归纳写入。
 */

/**
 * User Profile (1:1)
 *
 * user_profile 为自由文本 Markdown(用户在记忆管理页编辑),注入 user_agent,
 * 影响评判标准与 checklist 生成(≤ 2000 字符)。
 */
export interface UserPreferenceOut {
  /** 自由文本 Markdown(≤ 2000 字符) */
  user_profile: string
  /** agent 策略配置(检查点评估频率、打断权限等),null=未配置(用系统默认) */
  agent_policy: SaveAgentPolicyRequest | null
  /** 任务完成后是否自动生成练习题 draft(默认开;产出仍需预览确认) */
  auto_generate_practice: boolean
  /** 最后更新时间(ISO 字符串,未配置时为 null) */
  updated_at: string | null
}

/** 保存 User Profile 请求(PUT /memory/preferences body) */
export interface SaveUserPreferenceRequest {
  user_profile: string
}

/** 保存练习设置请求(PUT /memory/preferences/practice body) */
export interface SavePracticeSettingsRequest {
  auto_generate_practice: boolean
}

/**
 * agent 策略配置(PUT /memory/preferences/agent_policy body)
 *
 * 作为用户级默认值,任务级可通过 task.params["_agent_policy"] 覆盖。
 */
export interface SaveAgentPolicyRequest {
  /** 是否启用 user_agent(关闭=单 agent 模式,跳过评估/打断/验证) */
  user_agent_enabled: boolean
  /** user_agent 协作总轮次(1-10,仅 user_agent 启用时生效) */
  max_rounds: number
  /** 统一 K 值,每 K 个迭代评估一次 */
  checkpoint_interval: number
  /** 高级:内置 react_agent 专用 K 值(null=用统一值) */
  checkpoint_interval_builtin: number | null
  /** 高级:CLI agent 专用 K 值(null=用统一值) */
  checkpoint_interval_cli: number | null
  /** user_agent 是否能打断 react_agent */
  allow_interrupt: boolean
  /** 每轮最多打断次数(防死锁) */
  max_interrupts_per_round: number
  /** user_agent 是否能自己验证(实验性,先留开关) */
  allow_verify: boolean
  /** 验证授权默认模式:"direct" 直接执行 / "per_action" 逐动作授权(任务级可覆盖) */
  verifier_auth_mode_default: 'direct' | 'per_action'
  /** 执行智能体命令确认默认模式(任务级 _executor_command_confirm 可覆盖):
   *  - "always_approve":自动批准所有命令(注入 YOLO/never 策略,不弹窗)
   *  - "per_command":每个危险命令弹窗确认(CLI 走 ACP request_permission 通道) */
  executor_command_confirm_default: 'always_approve' | 'per_command'
}

/**
 * 全局长期记忆(1:1)
 *
 * 跨项目通用经验,注入 user_agent。自由文本,≤ 20000 字符。
 * 后端写入时还有合并截断(10000)+ 注入截断(2000)两道防线。
 */
export interface UserMemoryOut {
  content: string
  /** 最后更新时间(ISO 字符串,未配置时为 null) */
  updated_at: string | null
}

/** 保存全局长期记忆请求(PUT /memory/global body) */
export interface SaveUserMemoryRequest {
  content: string
}

/**
 * 系统级策略限制(GET /memory/policy-limits)
 *
 * 前端据此动态渲染输入上限,不硬编码。后端 max_rounds 可通过
 * 环境变量 AGENTPAIR_MAX_ROUNDS_LIMIT 调整。
 */
export interface PolicyLimitsOut {
  /** 协作总轮次上限(与后端 MAX_MAX_ROUNDS 对齐) */
  max_rounds: number
}

/**
 * 分项目记忆(1:N,按 repo_url 聚合)
 *
 * 项目由 orchestrator 在任务完成时自动归纳创建(_get_or_create_project),
 * 前端不提供"新建项目"入口,只能编辑/删除已由 agent 自动归纳产生的项目记录。
 */
export interface ProjectOut {
  /** 项目记忆 UUID(字符串) */
  id: string
  /** 归一化后的 repo_url,作为项目身份(不可改) */
  repo_url_normalized: string
  /** 原始 repo_url(展示用) */
  repo_url_raw: string
  /** 项目别名(用户可改,如 'AgentPair 主仓库') */
  alias: string | null
  /** 项目备注(用户可改,自由文本) */
  note: string | null
  /** 分项目记忆正文(注入 react_agent,≤ 20000 字符) */
  memory_content: string
  /** 上次自动归纳时间(ISO 字符串,可空) */
  last_summary_at: string | null
  created_at: string | null
  updated_at: string | null
}

/**
 * 保存分项目记忆请求(PUT /memory/projects/{id} body)
 *
 * 仅允许编辑 alias/note/memory_content(不修改 repo_url 与 last_summary_at)。
 */
export interface SaveProjectRequest {
  alias: string | null
  note: string | null
  memory_content: string
}

/** 分项目记忆列表响应(GET /memory/projects、DELETE 后响应) */
export interface ProjectListResponse {
  projects: ProjectOut[]
}

File: backend/app/schemas/memory.py
"""长期记忆管理相关的 Pydantic schema

对应 /memory 系列 API:
- UserPreferenceOut/SaveUserPreferenceRequest:User Profile (1:1)
- UserMemoryOut/SaveUserMemoryRequest:全局长期记忆(1:1)
- ProjectOut/SaveProjectRequest/ProjectListResponse:分项目记忆(1:N)

大小校验在 schema 层做第一道防线(写入时):
- user_profile ≤ 2000
- content(全局记忆) ≤ 20000
- memory_content(项目记忆) ≤ 20000

后续还有合并时截断(项目 8000 / 全局 10000)与注入时截断(2000)两道防线。
"""
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.models.practice import (
    LEARNING_TOPIC_ARCHITECTURE,
    LEARNING_TOPIC_CODING,
    LEARNING_TOPIC_SECURITY,
)


class UserPreferenceOut(BaseModel):
    """User Profile 响应(GET /memory/preferences)"""

    # 自由文本 Markdown(用户在记忆管理页编辑,注入 user_agent)
    user_profile: str = ""
    # agent 策略配置(检查点评估频率、打断权限、验证权限)
    # None 表示未配置(用系统默认),dict 表示用户自定义的覆盖值
    agent_policy: dict[str, Any] | None = None
    # 任务完成后是否自动生成练习题 draft(默认开)
    auto_generate_practice: bool = True
    # 当前学习主题(出题提示词按此切换,默认 security)
    learning_topic: str = "security"
    # 出题前沙箱已清理时是否重新 clone 恢复工作区(默认关)
    restore_workspace_for_practice: bool = False
    # 最后更新时间(可空 — 未配置时为 None;FastAPI 序列化为 ISO 字符串)
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class SaveUserPreferenceRequest(BaseModel):
    """保存 User Profile 请求(PUT /memory/preferences)"""

    user_profile: str = Field(default="", max_length=2000)


class SavePracticeSettingsRequest(BaseModel):
    """保存练习设置请求(PUT /memory/preferences/practice)

    - auto_generate_practice:任务完成后是否自动生成练习题 draft
      (产出仍需用户在预览对话框确认才转 active)
    - learning_topic:当前学习主题(security/architecture/coding),
      出题提示词按此切换;None 表示本次不修改
    - restore_workspace_for_practice:出题前沙箱已清理时是否重新 clone
      恢复工作区;None 表示本次不修改
    """

    auto_generate_practice: bool = True
    learning_topic: Literal[
        LEARNING_TOPIC_SECURITY,
        LEARNING_TOPIC_ARCHITECTURE,
        LEARNING_TOPIC_CODING,
    ] | None = None
    restore_workspace_for_practice: bool | None = None


class SaveAgentPolicyRequest(BaseModel):
    """保存 agent 策略配置请求(PUT /memory/preferences/agent_policy)

    结构与 agent_checkpoint.DEFAULT_AGENT_POLICY 对齐:
    - user_agent_enabled: 是否启用 user_agent(关闭=单 agent 模式)
    - max_rounds: user_agent 协作总轮次(上限由 MAX_MAX_ROUNDS 控制)
    - checkpoint_interval: 统一 K 值(每 K 个迭代评估一次)
    - checkpoint_interval_builtin: 内置 react_agent 专用 K 值(null=用统一值)
    - checkpoint_interval_cli: CLI agent 专用 K 值(null=用统一值)
    - allow_interrupt: user_agent 是否能打断 react_agent
    - max_interrupts_per_round: 每轮最多打断次数
    - allow_verify: user_agent 是否能调用 verifier_agent 验证(需任务配了 test_env_url)
    - verifier_auth_mode_default: 验证授权默认模式("direct"直接执行 / "per_action"逐动作授权)
    - executor_command_confirm_default: 执行智能体命令确认默认模式
        "always_approve" 自动批准所有命令 / "per_command" 每个危险命令弹窗确认
    """

    user_agent_enabled: bool = True
    # 上界在路由层用 MAX_MAX_ROUNDS 动态校验(schema 层只校验下界)
    max_rounds: int = Field(default=4, ge=1)
    checkpoint_interval: int = Field(default=10, ge=1, le=20)
    checkpoint_interval_builtin: int | None = Field(default=None, ge=1, le=20)
    checkpoint_interval_cli: int | None = Field(default=None, ge=1, le=20)
    allow_interrupt: bool = True
    max_interrupts_per_round: int = Field(default=2, ge=0, le=10)
    allow_verify: bool = False
    verifier_auth_mode_default: str = Field(default="per_action", pattern="^(direct|per_action)$")
    executor_command_confirm_default: str = Field(
        default="always_approve", pattern="^(always_approve|per_command)$"
    )


class PolicyLimitsOut(BaseModel):
    """系统级策略限制(GET /memory/policy-limits)

    前端据此动态渲染输入上限,不硬编码。后端 MAX_MAX_ROUNDS 可通过
    环境变量 AGENTPAIR_MAX_ROUNDS_LIMIT 调整。
    """

    max_rounds: int


class UserMemoryOut(BaseModel):
    """全局长期记忆响应(GET /memory/global)"""

    content: str = ""
    # 最后更新时间(可空 — 未配置时为 None;FastAPI 序列化为 ISO 字符串)
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class SaveUserMemoryRequest(BaseModel):
    """保存全局长期记忆请求(PUT /memory/global)"""

    content: str = Field(default="", max_length=20000)


class ProjectOut(BaseModel):
    """分项目记忆响应(GET /memory/projects、GET/PUT /memory/projects/{id})"""

    id: str
    repo_url_normalized: str
    repo_url_raw: str
    alias: str | None = None
    note: str | None = None
    memory_content: str = ""
    # 精简版记忆(系统生成,注入 system prompt 用;前端可查看)
    memory_summary: str = ""
    # 上次自动归纳时间(ISO 字符串,可空)
    last_summary_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None

    model_config = {"from_attributes": True}


class SaveProjectRequest(BaseModel):
    """保存分项目记忆请求(PUT /memory/projects/{id})

    仅允许编辑 alias/note/memory_content(不修改 repo_url 与 last_summary_at)。
    """

    alias: str | None = Field(default=None, max_length=255)
    note: str | None = Field(default=None)
    memory_content: str = Field(default="", max_length=20000)


class ProjectListResponse(BaseModel):
    """分项目记忆列表响应(GET /memory/projects、DELETE 后响应)"""

    projects: list[ProjectOut] = Field(default_factory=list)
